feature/aytochiloeches-36-access-to-participants-database

reperezz opened the pull request:
Close [#TG-ISSUE-36](https://tree.taiga.io/project/ailopez-chiloeches-decide/issue/36)

reperezz committed:
feature/aytochiloeches-36-access-to-participants-database

zjuanma commented on config/application_custom.rb:
```diff
@@ -1,3 +1,5 @@
+require 'csv'
```
Este require no seria mejor ponerlo donde se usa lo del csv, en custom/user.rb?

reperezz replied:
Sí, se puede añadir solo en el modelo User, pero estando en ese punto está accesible CSV en cualquier sitio de la aplicación.

ace replied:
+1 al comentario de @zjuanma, si en el futuro hace falta en más sitios, ya se verá.

eLafo replied:
+1

ace commented on app/controllers/custom/admin/participants_controller.rb:
```diff
@@ -0,0 +1,11 @@
+class Admin::ParticipantsController < Admin::BaseController
+
+  def index
+    @users = User.all.order('created_at desc')
```
El `order` debería ir en un scope del modelo, el controlador no necesita saber que la columna concreta se llama `created_at`.
El `all` es redundante, se está montando un objeto `relation` al que luego en el método `to_csv` se le va a volver a hacer un `all`.
Dado que no se hace nada con el objeto más allá de la acción, no es necesario que sea una variable de instancia, basta con `users`.

reperezz replied:
Ok

ace commented on app/models/custom/user.rb:
```diff
@@ -17,4 +17,15 @@ def level_two_verified?
+  def self.to_csv
+    attributes = %w{username email document_number document_type phone_number gender date_of_birth verified_at created_at}
+
+    CSV.generate(headers: true) do |csv|
```
¿Es necesario el `headers: true`? Estamos escribiendo el fichero, así que lo que le mandemos es lo que escribe, diría que esa opción solo tiene sentido en lectura.

reperezz replied:
Ese fichero es para comprobación por parte del cliente. El cliente lo va a leer y las cabeceras informan de lo que contiene cada columna. Es necesario.

ace replied:
¿Y eso que tiene que ver? Un .csv es un fichero de texto plano. Cuando lo lea la primera línea la interpretará como cabeceras y esa línea es la que metes con los `attributes`. Una cosa es que sean necesarias las cabeceras, otra que lo sea esa opción, que no debe hacer nada en escritura.

reperezz replied:
Ok, ya te comprendí. Cierto, ese parámetro solo vale en caso de lectura del csv. Lo elimino.

ace commented on app/views/custom/admin/_menu.html.erb:
```diff
@@ -0,0 +1,110 @@
+<nav class="admin-sidebar">
+  <ul id="admin_menu">
+
+    <li <%= "class=active" if controller_name == "tags" %>>
```
Esta misma construcción para asignar la clase css en 17 líneas pide a gritos un helper.

reperezz replied:
Si te fijas es un partial dentro de la carpeta custom que en el caso de Consul es una copia del partial definido en la rama principal de Consul. Podemos intentar seguir el modelo establecido por Consul o definir el nuestro propio, pero eso supondrá más dificultad a la hora hacer merges con nuevos contenidos en la rama master de Consul o esperas por nuestros PR a la línea oficial de Consul. En este proyecto habría que tener claro ese tema.

zjuanma replied:
no estoy deacuerdo con tigo @reperezz, desde el momento que haces un custom, eso ya no lo vas a volver a mezclar con master, así que puedes definir el helper como ha dicho @ace

reperezz replied:
Si hay cambios en el fichero que has customizado en la parte de master no añades esos cambios a tu fichero customizado? Entonces si hay nuevas entradas en ese menú de administración tú no las metes?

ace replied:
Sí, pero como es otro fichero y se van a tener que meter a mano, será el momento perfecto para adaptar esas nuevas entradas al helper.

reperezz replied:
Ok, añado un helper

ace replied:
Mirándolo con un poco más de atención, casi te diría que el helper puede montar no solo eso, sino toda la entrada de menú. Son todas idénticas salvo cuatro cosas, el nombre de controlador para la clase active, el path, el icono y el elemento de I18n. Con esos cuatro parámetros se pueden montar todos los elementos de menú en un solo método y cada uno te queda de una línea (salvo los que tienen un if condicionando).

reperezz committed:
[TG-ISSUE-36] Adjustments in access to participants database

shernade commented on app/views/custom/admin/verifications/index.html.erb:
```diff
@@ -11,7 +11,7 @@
       </td>
       <td><%= user.document_number %></td>
-      <td><%= I18n.l(user.date_of_birth.to_datetime.to_date) %></td>
+      <td><%= I18n.l(user.date_of_birth.to_datetime.to_date) if user.date_of_birth %></td>
```
el .to_datetime.to_date es necesario?

ace replied:
Yo no le veo ningún sentido, `date_of_birth` (empezamos bien con el nombre) es un `datetime` (toma moreno) en bbdd, así que el `.to_datetime` parece que no hace ninguna falta. (Y ya se que todo lo anterior no es cosa nuestra, que venía así, pero si no lo digo reviento :D)

reperezz replied:
Yo solo le he añadido el if al final porque si no tenía valor en el campo date_of_birth daba error. Pero es cierto que al ser un datetime no es necesario el .to_datetime. El .to_date si para mostrar solo la fecha en formato dd/mm/YYYY

eLafo replied:
Pues si el to_datetime no es necesario, quitémoslo...

ace replied:
Como nota al margen, siendo estrictos el `to_date` tampoco haría falta, si solo se quiere la fecha, bastaría con pasarle al `I18n.l` el formato que se quiere.

eLafo reviewed (changes requested):
Efectivamente hay que quitar la opción headers del csv y extraer el código que decide si se añade la clase "active" a un helper

reperezz committed:
[TG-ISSUE-36] Adjustments in access to participants database

ace commented on app/helpers/tags_helper.rb:
```diff
@@ -11,4 +11,21 @@ def taggable_path(taggable_type, tag_name)
     end
   end
 
+  def nav_element(match_controller, path = root_path, legend = '', options = {}, &block)
```
Me sobra un poco el baile con las `options` :tag_name y :class_icon. Una nunca se pasa y siempre vale el valor por defecto y la otra siempre se pasa y el valor por defecto no pinta nada. Y sobre todo el bloque, no tenemos ese uso por ningún lado.

eLafo replied:
Yo creo que tiene sentido hacer configurable tanto el tag como el class icon. Otra cosa es que ahora mismo no lo utilicemos como cliente y que hayamos pecado un pelín de sobrearquitectura, pero a mí personalmente me cuadra.

ace replied:
El tag no lo veo, toda esa navegación va en un `ul`, si no va a ser un `li` el cambio va a ser más grande que el helper, pero bueno, tampoco tiene más importancia. Del class icon lo que comento es que siempre se está recibiendo, no le veo sentido al || con un valor por defecto (que además me parece arbitrario, ni siquiera es el icono más usado). Lo del bloque sí que ya me parece demasiado sobrearquitecturado.

eLafo replied:
Bueno, supongo que es un tema de dónde poner la línea. A mí este helper se me parece bastante  los que te da rails, con opciones para configurar diferentes cosas. Por eso a mí no me canta tanto. Me parece que aunque sea una solución relativamente grande para un caso puntual, se compensa siguiendo cosas que hace rails y por lo tanto respetando el principio de la mínima sorpresa.

Dicho esto a mí ambas opciones me parecen válidas con sus pros y sus contras. Puesto que las consecuencias de la decisión no me las como yo, sino que os las coméis los que estáis en el día a día del proyecto, lo suyo es que llegués a un consenso ;-)

reperezz replied:
Yo ya no perdería más tiempo con esto. Esas opciones que a Andrés le parecen sobrearquitecturación, palabra sobrecargada en sí misma, dan flexibilidad al método y que pueda usarse el día de mañana en una pestañas o con un contenido diferente, sin que suponga una difícil comprensión de lo que ahí se está haciendo o de rendimiento.

ace replied:
Hombre, algo más complicado lo ha hecho. Como no hay teses te has ahorrado esa parte, pero esta flexibilidad llevaría asociados varios teses adicionales. Que un helper de rails te de esta flexibilidad me parece bien, el framework te ofrece posibilidades para usarlo como quieras/necesites. En este caso, teníamos todos los usos antes del helper, añadir 'futuribles' entra en esa línea de la optimización prematura que quizá no se use nunca.
En cualquier caso, tienes razón, ya le hemos dedicado bastante tiempo, por mi adelante.

eLafo replied:
+1 a lo de los tests que comenta @ace

eLafo commented:
Rubocop se queja en varios de los archivos modificados. Antes de mergear, hay que arreglar esas ofensas, o al menos justificarlas

eLafo reviewed (changes requested):
Rubocop se queja en los siguientes archivos:

- app/controllers/custom/admin/participants_controller.rb
- app/models/custom/user.rb
- app/views/custom/admin/_menu.html.erb
- app/views/custom/admin/verifications/index.html.erb
- config/locales/custom/admin.es.yml
- config/routes.rb

Antes de mergear, hay que corregir esas ofensas, o al menos justificarlas

eLafo committed:
rubocop configuration

Upgraded rubocop gem to 0.48.1
Regenerated .rubocop_todo.yml
Created .rubocop_custom.yml where every custom configuration should be
.rubocop.yml inherits from .rubocop_custom.yml and .rubocop_todo.yml
Fixed conditional offense in app/controllers/valuation/spending_proposals_controller.rb

eLafo committed:
removed inheritance in .rubocop_custom.yml

reperezz committed:
Merge pull request #3 from aspgems/TR/rubocop-config

rubocop configuration

reperezz committed:
Merge branch 'master' into feature/aytochiloeches-36-access-to-participants-database

reperezz committed:
Updating rubocop_todo and removing offense

reperezz committed:
Rubocop configuration for Aytochiloeches

reperezz committed:
Merge branch with previous changes in repo
